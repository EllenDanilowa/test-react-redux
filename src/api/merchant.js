import _merchants from './data/merchants.json';
import _bids from './data/bids.json';
import {
  getRandomId,
  getRandomInt,
  convertImage
} from './merchant.utils';

const TIMEOUT = 250;

/**
 * Mocking client-server processing
 */
const _getAll = () => {
  return new Promise((resolve) => {
    const data = {};

    data.items = _merchants.map((item) => {
      item.bids = _bids.slice(0, getRandomInt(_bids.length));

      return item;
    });

    setTimeout(() => resolve({data}), TIMEOUT);
  });
};

const _update = (merchant) => {
  return new Promise((resolve) => {
    const newMerchant = {
      ...merchant,
      id: merchant.id || getRandomId(),
      bids: merchant.id ? merchant.bids : _bids.slice(0, getRandomInt(_bids.length)),
      avatar: null
    };
    const resolvePromise = () => setTimeout(() => resolve({data: newMerchant}), TIMEOUT);

    if (merchant.avatar && merchant.avatar.length) {
      convertImage(merchant.avatar[0]).then((image) => {
        newMerchant.avatarUrl = image;

        resolvePromise();
      }).catch((error) => {
        console.log(error);

        resolvePromise();
      });
    } else {
      resolvePromise();
    }
  });
};

const _delete = (id) => {
  return new Promise((resolve) => {
    setTimeout(() => resolve({data: id}), TIMEOUT);
  });
};

export default {
  getAll: _getAll,
  create: _update,
  update: _update,
  delete: _delete
};
