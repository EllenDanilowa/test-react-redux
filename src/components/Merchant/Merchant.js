import React, {Component} from 'react';
import {Link} from 'react-router-dom';
import PropTypes from 'prop-types';
import {
  MerchantWrapper,
  Avatar,
  Name,
  ContentWrapper,
  Icon,
  LargeIcon,
  PremiumIcon,
  ActionsWrapper,
  ContentItem,
  DeleteButton
} from './Merchant.styled';
import Bid from './Bid/Bid';
import Accordion from './Accordion/Accordion';
import defaultAvatar from './assets/default-avatar.png';
import mailIcon from './assets/mail.svg';
import phoneIcon from './assets/phone.svg';
import premiumIcon from './assets/premium-user.svg';
import editIcon from './assets/edit.svg';
import deleteIcon from './assets/delete.svg';

class Merchant extends Component {
  constructor(props) {
    super(props);

    this.deleteMerchant = this.deleteMerchant.bind(this);
  }

  deleteMerchant() {
    this.props.delete(this.props.item.id);
  }

  render() {
    const {item} = this.props;

    return (
      <MerchantWrapper>
        {item.hasPremium && <PremiumIcon src={premiumIcon} alt="Premium user" title="Premium user"/>}
        <Avatar src={item.avatarUrl || defaultAvatar} alt="avatar" />
        <ContentWrapper>
          <Name>{item.firstname} {item.lastname}</Name>

          <ContentItem>
            <Icon src={mailIcon}/>
            <span>{item.email}</span>
          </ContentItem>
          <ContentItem>
            <Icon src={phoneIcon}/>
            <span>{item.phone}</span>
          </ContentItem>

          {Boolean(item.bids.length) && (
            <Accordion title="See bids">
              <div>{item.bids.map((bid) => (
                <Bid key={bid.id} item={bid}/>
              ))}</div>
            </Accordion>
          )}

        </ContentWrapper>
        <ActionsWrapper>
          <Link to={`/edit/${item.id}`}>
            <LargeIcon src={editIcon} />
          </Link>
          <DeleteButton onClick={this.deleteMerchant}>
            <LargeIcon src={deleteIcon} />
          </DeleteButton>
        </ActionsWrapper>
      </MerchantWrapper>
    );
  }
}

Merchant.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string.isRequired,
    firstname: PropTypes.string.isRequired,
    lastname: PropTypes.string.isRequired,
    avatarUrl: PropTypes.string,
    email: PropTypes.string.isRequired,
    phone: PropTypes.string.isRequired,
    hasPremium: PropTypes.bool.isRequired,
    bids: PropTypes.array
  }),
  delete: PropTypes.func.isRequired
};

export default Merchant;
