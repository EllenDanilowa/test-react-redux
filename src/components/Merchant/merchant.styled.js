import styled from 'styled-components';
import {Link} from 'react-router-dom';
import colors from '../../styles/colors';


const borderColor = '#cccccc';

export const MerchantWrapper = styled.div`
  border: 1px solid ${borderColor};
  border-radius: 4px;
  color: ${colors.title};
  display: flex;
  margin: 12px 0;
  padding: 12px;
  overflow: hidden;
  position: relative;
`;

const avatarSize = '130px';
export const Avatar = styled.img`
  align-self: flex-start;
  flex: 0 0 ${avatarSize};
  height: auto;
  width: ${avatarSize};
`;

export const Name = styled.h4`
  font-size: 20px;
  font-weight: 500;
  margin: 12px 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

export const Icon = styled.img`
  padding-right: 4px;
  width: 15px;
`;

export const ContentWrapper = styled.div`
  flex: 1;
  min-width: 0;
  padding: 16px;
`;

export const ContentItem = styled.div`
  font-size: 14px;
  display: flex;
  margin: 7px 0;
`;

export const PremiumIcon = styled.img`
  position: absolute;
  right: 12px;
  top: 12px;
  width: 30px;
`;

export const ActionsWrapper = styled.div`
  align-self: flex-end;
  display: flex;
  flex: 0 0 60px;
  justify-content: flex-end;
`;

export const ActionIcon = styled.img`
  width: 20px;
`;

export const ActionLink = styled(Link)`
  display: block;
`;

export const DeleteButton = styled.button`
  border: none;
  background: none;
  cursor: pointer;
  outline: 0;
  padding: 0;
  margin-left: 8px;
`;
