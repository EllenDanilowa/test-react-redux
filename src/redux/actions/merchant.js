import API from '../../api/merchant';
import { 
    FETCH_MERCHANTS_BEGIN, 
    FETCH_MERCHANTS_SUCCESS, 
    FETCH_MERCHANTS_FAILURE
} from '../types/merchant';

export const fetchMerchants = () => {
    return dispatch => {
      dispatch(fetchMerchantsBegin());
      return API.getAll()
        .then(json => {
          dispatch(fetchMerchantsSuccess(json.data));
          return json.data;
        })
        .catch(error => dispatch(fetchMerchantsFailure(error)));
    };
  };

export const fetchMerchantsBegin = () => ({
  type: FETCH_MERCHANTS_BEGIN
});

export const fetchMerchantsSuccess = merchants => ({
  type: FETCH_MERCHANTS_SUCCESS,
  payload: { merchants }
});

export const fetchMerchantsFailure = error => ({
  type: FETCH_MERCHANTS_FAILURE,
  payload: { error }
});