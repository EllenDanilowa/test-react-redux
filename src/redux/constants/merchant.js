
export const FETCH_MERCHANTS_BEGIN   = 'FETCH_MERCHANTS_BEGIN';
export const FETCH_MERCHANTS_SUCCESS = 'FETCH_MERCHANTS_SUCCESS';
export const FETCH_MERCHANTS_FAILURE = 'FETCH_MERCHANTS_FAILURE';

export const ADD_NEW_MERCHANT = 'ADD_NEW_MERCHANT';
export const EDIT_MERCHANT = 'EDIT_MERCHANT';
export const DELETE_MERCHANT = 'DELETE_MERCHANT';