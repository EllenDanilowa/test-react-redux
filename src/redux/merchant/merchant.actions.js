import API from '../../../api/merchant';
import {push} from 'connected-react-router';
import {
  FETCH_MERCHANTS_BEGIN,
  FETCH_MERCHANTS_SUCCESS,
  FETCH_MERCHANTS_FAILURE,
  UPDATE_VISIBLE_MERCHANTS,
  CREATE_NEW_MERCHANT_SUCCESS,
  UPDATE_MERCHANT_SUCCESS,
  DELETE_MERCHANT_SUCCESS,
} from './merchant.constants';

export const fetchMerchants = () => {
  return (dispatch) => {
    dispatch(fetchMerchantsBegin());

    return API.getAll()
      .then((json) => dispatch(fetchMerchantsSuccess(json.data)))
      .catch((error) => dispatch(fetchMerchantsFailure(error)));
  };
};

export const createMerchant = (item) => {
  return (dispatch) => {
    //dispatch(createMerchantsBegin());

    return API.create(item)
      .then((json) => {
        dispatch(createNewMerchantSuccess((json.data)));
        dispatch(push(''));
      });
      //.catch((error) => dispatch(createMerchantFailure(error)));
  };
};

export const updateMerchant = (item) => {
  return (dispatch) => {
    //dispatch(updateMerchantsBegin());

    return API.update(item)
      .then((json) => {
        dispatch(updateMerchantSuccess((json.data)));
        dispatch(push(''));
      });
    //.catch((error) => dispatch(updateMerchantsBegin(error)));
  };
};

export const deleteMerchant = (item) => {
  return (dispatch) => {
    //dispatch(deleteMerchantsBegin());

    return API.delete(item)
      .then((json) => {
        dispatch(deleteMerchantSuccess((json.data)));
        dispatch(push(''));
      });
    //.catch((error) => dispatch(deleteMerchantsBegin(error)));
  };
};

export const fetchMerchantsBegin = () => ({
  type: FETCH_MERCHANTS_BEGIN
});

export const fetchMerchantsSuccess = (data) => ({
  type: FETCH_MERCHANTS_SUCCESS,
  payload: {
    merchants: data.items
  }
});

export const fetchMerchantsFailure = (error) => ({
  type: FETCH_MERCHANTS_FAILURE,
  payload: {error}
});

export const updateVisibleMerchants = (from, to) => ({
  type: UPDATE_VISIBLE_MERCHANTS,
  payload: {from, to}
});

export const createNewMerchantSuccess = (merchant) => ({
  type: CREATE_NEW_MERCHANT_SUCCESS,
  payload: {merchant}
});

export const updateMerchantSuccess = (merchant) => ({
  type: UPDATE_MERCHANT_SUCCESS,
  payload: {merchant}
});

export const deleteMerchantSuccess = (id) => ({
  type: DELETE_MERCHANT_SUCCESS,
  payload: {id}
});
