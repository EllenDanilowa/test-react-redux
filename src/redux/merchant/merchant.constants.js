
export const FETCH_MERCHANTS_BEGIN   = 'FETCH_MERCHANTS_BEGIN';
export const FETCH_MERCHANTS_SUCCESS = 'FETCH_MERCHANTS_SUCCESS';
export const FETCH_MERCHANTS_FAILURE = 'FETCH_MERCHANTS_FAILURE';

export const CREATE_NEW_MERCHANT_SUCCESS = 'CREATE_NEW_MERCHANT_SUCCESS';

export const UPDATE_MERCHANT_SUCCESS = 'UPDATE_MERCHANT_SUCCESS';
export const DELETE_MERCHANT_SUCCESS = 'DELETE_MERCHANT_SUCCESS';
