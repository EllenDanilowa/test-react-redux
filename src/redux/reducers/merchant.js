import {
    FETCH_MERCHANTS_BEGIN,
    FETCH_MERCHANTS_SUCCESS,
    FETCH_MERCHANTS_FAILURE
} from '../types/merchant';

const initialState = {
    items: [],
    loading: false,
    error: null
  };
  
export default (state = {}, action) => {
    switch (action.type) {
        case FETCH_MERCHANTS_BEGIN:
            return {
                ...state,
                loading: true,
                error: null
            };
        case FETCH_MERCHANTS_SUCCESS: {
            return {
                ...state,
                loading: false,
                items: action.payload.merchants
            }
        }
        case FETCH_MERCHANTS_FAILURE:
            return {
                ...state,
                loading: false,
                error: action.payload.error,
                items: []
            };
        default:
            return state;
    }
};