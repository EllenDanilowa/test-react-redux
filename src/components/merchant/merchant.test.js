import React from 'react';
import {MemoryRouter} from 'react-router';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import Merchant from './merchant';
import {DeleteButton} from './merchant.styled';

jest.mock('./bid/bid', () => () => 'Bid component');
jest.mock('./accordion/accordion', () => ({children}) => (<div><p>Accordion</p><div>'{children}'</div></div>));

describe('Merchant', () => {
  let element;
  let item;
  let deleteFunc;

  const createElement = ({deleteFunc, ...props}, isSnapshot = true) => {
    const template = (<MemoryRouter><Merchant delete={deleteFunc} {...props} /></MemoryRouter>);

    return isSnapshot ?
      renderer.create(template).toJSON() :
      mount(template);
  };

  beforeEach(() => {
    item = {
      id: '1',
      firstname: 'name',
      lastname: 'lastname',
      avatarUrl: '',
      email: 'test@test.com',
      phone: '123456789',
      hasPremium: false,
      bids: []
    };
    deleteFunc = jest.fn();
  });

  it('renders merchant with bids', () => {
    item.bids.push({
      id: '1',
      carTitle: 'title',
      amount: 2500,
      created: 1595116800000
    });

    element = createElement({item, deleteFunc});

    expect(element).toMatchSnapshot();
  });

  it('renders merchant without bids', () => {
    element = createElement({item, deleteFunc});

    expect(element).toMatchSnapshot();
  });

  it('renders premium merchant', () => {
    item.hasPremium = true;
    element = createElement({item, deleteFunc});

    expect(element).toMatchSnapshot();
  });

  it('triggers delete callback if delete button is clicked', () => {
    element = createElement({item, deleteFunc}, false);

    element.find(DeleteButton).simulate('click');
    expect(deleteFunc).toHaveBeenCalledWith(item.id);
  });
});
