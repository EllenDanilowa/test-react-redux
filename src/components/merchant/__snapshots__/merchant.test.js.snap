// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merchant renders merchant with bids 1`] = `
<div
  className="sc-AxjAm htNEyp"
>
  <img
    alt="avatar"
    className="sc-AxirZ fmnQqz"
    src="default-avatar.png"
  />
  <div
    className="sc-AxhUy lifGgf"
  >
    <h4
      className="sc-AxiKw erQByv"
    >
      name
       
      lastname
    </h4>
    <div
      className="sc-AxgMl bQvVVa"
    >
      <img
        className="sc-AxhCb gnNDWi"
        src="mail.svg"
      />
      <span>
        test@test.com
      </span>
    </div>
    <div
      className="sc-AxgMl bQvVVa"
    >
      <img
        className="sc-AxhCb gnNDWi"
        src="phone.svg"
      />
      <span>
        123456789
      </span>
    </div>
    <div>
      <p>
        Accordion
      </p>
      <div>
        '
        <div>
          Bid component
        </div>
        '
      </div>
    </div>
  </div>
  <div
    className="sc-Axmtr bbWZXo"
  >
    <a
      className="sc-fzozJi jEPYlu"
      href="/update/1"
      onClick={[Function]}
    >
      <img
        className="sc-AxmLO kRZbDE"
        src="edit.svg"
      />
    </a>
    <button
      className="sc-fzoLsD dcjTmq"
      onClick={[Function]}
    >
      <img
        className="sc-AxmLO kRZbDE"
        src="delete.svg"
      />
    </button>
  </div>
</div>
`;

exports[`Merchant renders merchant without bids 1`] = `
<div
  className="sc-AxjAm htNEyp"
>
  <img
    alt="avatar"
    className="sc-AxirZ fmnQqz"
    src="default-avatar.png"
  />
  <div
    className="sc-AxhUy lifGgf"
  >
    <h4
      className="sc-AxiKw erQByv"
    >
      name
       
      lastname
    </h4>
    <div
      className="sc-AxgMl bQvVVa"
    >
      <img
        className="sc-AxhCb gnNDWi"
        src="mail.svg"
      />
      <span>
        test@test.com
      </span>
    </div>
    <div
      className="sc-AxgMl bQvVVa"
    >
      <img
        className="sc-AxhCb gnNDWi"
        src="phone.svg"
      />
      <span>
        123456789
      </span>
    </div>
  </div>
  <div
    className="sc-Axmtr bbWZXo"
  >
    <a
      className="sc-fzozJi jEPYlu"
      href="/update/1"
      onClick={[Function]}
    >
      <img
        className="sc-AxmLO kRZbDE"
        src="edit.svg"
      />
    </a>
    <button
      className="sc-fzoLsD dcjTmq"
      onClick={[Function]}
    >
      <img
        className="sc-AxmLO kRZbDE"
        src="delete.svg"
      />
    </button>
  </div>
</div>
`;

exports[`Merchant renders premium merchant 1`] = `
<div
  className="sc-AxjAm htNEyp"
>
  <img
    alt="Premium user"
    className="sc-AxheI jNJVop"
    src="premium-user.svg"
    title="Premium user"
  />
  <img
    alt="avatar"
    className="sc-AxirZ fmnQqz"
    src="default-avatar.png"
  />
  <div
    className="sc-AxhUy lifGgf"
  >
    <h4
      className="sc-AxiKw erQByv"
    >
      name
       
      lastname
    </h4>
    <div
      className="sc-AxgMl bQvVVa"
    >
      <img
        className="sc-AxhCb gnNDWi"
        src="mail.svg"
      />
      <span>
        test@test.com
      </span>
    </div>
    <div
      className="sc-AxgMl bQvVVa"
    >
      <img
        className="sc-AxhCb gnNDWi"
        src="phone.svg"
      />
      <span>
        123456789
      </span>
    </div>
  </div>
  <div
    className="sc-Axmtr bbWZXo"
  >
    <a
      className="sc-fzozJi jEPYlu"
      href="/update/1"
      onClick={[Function]}
    >
      <img
        className="sc-AxmLO kRZbDE"
        src="edit.svg"
      />
    </a>
    <button
      className="sc-fzoLsD dcjTmq"
      onClick={[Function]}
    >
      <img
        className="sc-AxmLO kRZbDE"
        src="delete.svg"
      />
    </button>
  </div>
</div>
`;
